OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp

```diff
@@ -0,0 +1,223 @@
+  // move chain on for next layer
+  layer_instance_create_info->u.pLayerInfo = layer_instance_create_info->u.pLayerInfo->pNext;
+
+  auto create_instance =
+      absl::bit_cast<PFN_vkCreateInstance>(gpa(VK_NULL_HANDLE, "vkCreateInstance"));
+
+  VkResult result = create_instance(instance_create_info, allocator, instance);
+
+  {
+    absl::WriterMutexLock lock(&layer_mutex);
+    dispatch_table.CreateInstanceDispatchTable(*instance, gpa);
+  }
+
+  return result;
+}
+
+VK_LAYER_EXPORT void VKAPI_CALL
+OrbitCaptureClientDestroyInstance(VkInstance instance, const VkAllocationCallbacks* /*allocator*/) {
+  absl::WriterMutexLock lock(&layer_mutex);
+  dispatch_table.DestroyInstance(instance);
+}
+
+VK_LAYER_EXPORT VkResult VKAPI_CALL OrbitCaptureClientCreateDevice(
+    VkPhysicalDevice physical_device, const VkDeviceCreateInfo* device_create_info,
+    const VkAllocationCallbacks* allocator, VkDevice* device) {
+  auto* layer_device_create_info =
+      absl::bit_cast<VkLayerDeviceCreateInfo*>(device_create_info->pNext);
+
+  // step through the chain of pNext until we get to the link info
+  while (layer_device_create_info != nullptr &&
+         (layer_device_create_info->sType != VK_STRUCTURE_TYPE_LOADER_DEVICE_CREATE_INFO ||
+          layer_device_create_info->function != VK_LAYER_LINK_INFO)) {
+    layer_device_create_info =
+        absl::bit_cast<VkLayerDeviceCreateInfo*>(layer_device_create_info->pNext);
+  }
+
+  if (layer_device_create_info == nullptr) {
+    // No loader instance create info
+    return VK_ERROR_INITIALIZATION_FAILED;
+  }
+
+  PFN_vkGetInstanceProcAddr gipa =
+      layer_device_create_info->u.pLayerInfo->pfnNextGetInstanceProcAddr;
+  PFN_vkGetDeviceProcAddr gdpa = layer_device_create_info->u.pLayerInfo->pfnNextGetDeviceProcAddr;
+  // move chain on for next layer
+  layer_device_create_info->u.pLayerInfo = layer_device_create_info->u.pLayerInfo->pNext;
+
+  auto create_device = absl::bit_cast<PFN_vkCreateDevice>(gipa(VK_NULL_HANDLE, "vkCreateDevice"));
+
+  VkResult result = create_device(physical_device, device_create_info, allocator, device);
+
+  {
+    absl::WriterMutexLock lock(&layer_mutex);
+    dispatch_table.CreateDeviceDispatchTable(*device, gdpa);
+  }
+
+  return result;
+}
+
+VK_LAYER_EXPORT void VKAPI_CALL
+OrbitCaptureClientDestroyDevice(VkDevice device, const VkAllocationCallbacks* /*allocator*/) {
+  absl::WriterMutexLock lock(&layer_mutex);
+  dispatch_table.DestroyDevice(device);
+}
+
+// --------------------------------------------------------------------------------
+// Enumeration function
+// --------------------------------------------------------------------------------
+
+VK_LAYER_EXPORT VkResult VKAPI_CALL OrbitCaptureClientEnumerateInstanceLayerProperties(
+    uint32_t* property_count, VkLayerProperties* properties) {
+  if (property_count != nullptr) {
+    *property_count = 1;
+  }
+
+  if (properties != nullptr) {
+    strcpy(properties->layerName, kLayerName);
```

Comment:
I am not sure about `strcpy` vs. `snprintf` @danielfenner or @ronaldfw maybe you do have some insights?

Reply:
I did some investigation about this and it seems _snprintf_ is safer and recommended over strcpy so I have updated the code. Thanks!

Reply:
This seems to be breaking the build and it is related to the size; I am using `strlen `which should be returning the correct size? What should I do instead?

Reply:
I am getting the same on my version. Haven't had the chance to look into this further. From my side, strcpy would be also fine. I'd be curious about  @danielfenner or @ronaldfw point of view.

Reply:
The second parameter can be 256 (the size of properties->layerName) here for this to be correct. The issue that you see is that strlen does not count the null termination character but that also needs to be copied. 

I would also suggest to use the Abseil version of snprintf: https://github.com/abseil/abseil-cpp/blob/e96d49687d9c078f2d47356b6723c3b5715493f7/absl/strings/str_format.h#L423

Reply:
Updated, thanks!

Reply:
Latest version looks good to me!